### PR TITLE
mrustc: 0.9 -> 0.10

### PR DESCRIPTION
**Pull request by progval**
###### Description of changes

And switch from Rust 1.29 to 1.54, as this is the top version supported by mrustc.

Addresses https://github.com/NixOS/nixpkgs/issues/44372#issuecomment-1497723466 (cc @RaitoBezarius)

This fails to build on my machine, with this error:

```
(171/175) BUILDING rustc_privacy v0.0.0
> /nix/store/54w5f18vb7xy2frx5i091q77l2zaz76x-mrustc-0.10/bin/mrustc rustc-1.54.0-src/compiler/rustc_privacy/src/lib.rs -o output-1.54.0/rustc-build/librustc_privacy.rlib --crate-name rustc_privacy --crate-type rlib -C emit-depfile=output-1.54.0/rustc-build/librustc_privacy.rlib.d --crate-tag 0_0_0 -g --cfg debug_assertions -O -L output-1.54.0 -L output-1.54.0/rustc-build --edition 2018 --extern rustc_middle=output-1.54.0/rustc-build/librustc_middle.rlib --extern rustc_attr=output-1.54.0/rustc-build/librustc_attr.rlib --extern rustc_errors=output-1.54.0/rustc-build/librustc_errors.rlib --extern rustc_hir=output-1.54.0/rustc-build/librustc_hir.rlib --extern rustc_typeck=output-1.54.0/rustc-build/librustc_typeck.rlib --extern rustc_session=output-1.54.0/rustc-build/librustc_session.rlib --extern rustc_span=output-1.54.0/rustc-build/librustc_span.rlib --extern rustc_data_structures=output-1.54.0/rustc-build/librustc_data_structures.rlib --extern rustc_trait_selection=output-1.54.0/rustc-build/librustc_trait_selection.rlib --extern tracing=output-1.54.0/rustc-build/libtracing-0_1_26_H30002.rlib
(172/175) BUILDING rustc_interface v0.0.0 with features [llvm,rustc_codegen_llvm]
> /nix/store/54w5f18vb7xy2frx5i091q77l2zaz76x-mrustc-0.10/bin/mrustc rustc-1.54.0-src/compiler/rustc_interface/src/lib.rs -o output-1.54.0/rustc-build/librustc_interface.rlib --crate-name rustc_interface --crate-type rlib -C emit-depfile=output-1.54.0/rustc-build/librustc_interface.rlib.d --crate-tag 0_0_0_H3 -g --cfg debug_assertions -O -L output-1.54.0 -L output-1.54.0/rustc-build --cfg feature="llvm" --cfg feature="rustc_codegen_llvm" --edition 2018 --extern libc=output-1.54.0/rustc-build/liblibc-0_2_95_H20.rlib --extern tracing=output-1.54.0/rustc-build/libtracing-0_1_26_H30002.rlib --extern rustc_rayon_core=output-1.54.0/rustc-build/librustc_rayon_core-0_3_1.rlib --extern rayon=output-1.54.0/rustc-build/librustc_rayon-0_3_1.rlib --extern smallvec=output-1.54.0/rustc-build/libsmallvec-1_6_1_H12.rlib --extern rustc_ast=output-1.54.0/rustc-build/librustc_ast.rlib --extern rustc_attr=output-1.54.0/rustc-build/librustc_attr.rlib --extern rustc_builtin_macros=output-1.54.0/rustc-build/librustc_builtin_macros.rlib --extern rustc_expand=output-1.54.0/rustc-build/librustc_expand.rlib --extern rustc_parse=output-1.54.0/rustc-build/librustc_parse.rlib --extern rustc_session=output-1.54.0/rustc-build/librustc_session.rlib --extern rustc_span=output-1.54.0/rustc-build/librustc_span.rlib --extern rustc_serialize=output-1.54.0/rustc-build/librustc_serialize.rlib --extern rustc_middle=output-1.54.0/rustc-build/librustc_middle.rlib --extern rustc_ast_lowering=output-1.54.0/rustc-build/librustc_ast_lowering.rlib --extern rustc_ast_passes=output-1.54.0/rustc-build/librustc_ast_passes.rlib --extern rustc_incremental=output-1.54.0/rustc-build/librustc_incremental.rlib --extern rustc_traits=output-1.54.0/rustc-build/librustc_traits.rlib --extern rustc_data_structures=output-1.54.0/rustc-build/librustc_data_structures.rlib --extern rustc_codegen_ssa=output-1.54.0/rustc-build/librustc_codegen_ssa.rlib --extern rustc_symbol_mangling=output-1.54.0/rustc-build/librustc_symbol_mangling.rlib --extern rustc_codegen_llvm=output-1.54.0/rustc-build/librustc_codegen_llvm.rlib --extern rustc_hir=output-1.54.0/rustc-build/librustc_hir.rlib --extern rustc_index=output-1.54.0/rustc-build/librustc_index.rlib --extern rustc_metadata=output-1.54.0/rustc-build/librustc_metadata.rlib --extern rustc_mir=output-1.54.0/rustc-build/librustc_mir.rlib --extern rustc_mir_build=output-1.54.0/rustc-build/librustc_mir_build.rlib --extern rustc_passes=output-1.54.0/rustc-build/librustc_passes.rlib --extern rustc_typeck=output-1.54.0/rustc-build/librustc_typeck.rlib --extern rustc_lint=output-1.54.0/rustc-build/librustc_lint.rlib --extern rustc_errors=output-1.54.0/rustc-build/librustc_errors.rlib --extern rustc_plugin_impl=output-1.54.0/rustc-build/librustc_plugin_impl.rlib --extern rustc_privacy=output-1.54.0/rustc-build/librustc_privacy.rlib --extern rustc_query_impl=output-1.54.0/rustc-build/librustc_query_impl.rlib --extern rustc_resolve=output-1.54.0/rustc-build/librustc_resolve.rlib --extern rustc_trait_selection=output-1.54.0/rustc-build/librustc_trait_selection.rlib --extern rustc_ty_utils=output-1.54.0/rustc-build/librustc_ty_utils.rlib --extern tempfile=output-1.54.0/rustc-build/libtempfile-3_2_0.rlib
(173/175) BUILDING rustc_driver v0.0.0 with features [llvm]
> /nix/store/54w5f18vb7xy2frx5i091q77l2zaz76x-mrustc-0.10/bin/mrustc rustc-1.54.0-src/compiler/rustc_driver/src/lib.rs -o output-1.54.0/rustc-build/librustc_driver.rlib --crate-name rustc_driver --crate-type rlib -C emit-depfile=output-1.54.0/rustc-build/librustc_driver.rlib.d --crate-tag 0_0_0_H1 -g --cfg debug_assertions -O -L output-1.54.0 -L output-1.54.0/rustc-build --cfg feature="llvm" --edition 2018 --extern libc=output-1.54.0/rustc-build/liblibc-0_2_95_H20.rlib --extern atty=output-1.54.0/rustc-build/libatty-0_2_14.rlib --extern tracing=output-1.54.0/rustc-build/libtracing-0_1_26_H30002.rlib --extern tracing_subscriber=output-1.54.0/rustc-build/libtracing_subscriber-0_2_18_H7ffff.rlib --extern tracing_tree=output-1.54.0/rustc-build/libtracing_tree-0_1_9_H1.rlib --extern rustc_middle=output-1.54.0/rustc-build/librustc_middle.rlib --extern rustc_ast_pretty=output-1.54.0/rustc-build/librustc_ast_pretty.rlib --extern rustc_target=output-1.54.0/rustc-build/librustc_target.rlib --extern rustc_lint=output-1.54.0/rustc-build/librustc_lint.rlib --extern rustc_data_structures=output-1.54.0/rustc-build/librustc_data_structures.rlib --extern rustc_errors=output-1.54.0/rustc-build/librustc_errors.rlib --extern rustc_feature=output-1.54.0/rustc-build/librustc_feature.rlib --extern rustc_hir=output-1.54.0/rustc-build/librustc_hir.rlib --extern rustc_hir_pretty=output-1.54.0/rustc-build/librustc_hir_pretty.rlib --extern rustc_metadata=output-1.54.0/rustc-build/librustc_metadata.rlib --extern rustc_mir=output-1.54.0/rustc-build/librustc_mir.rlib --extern rustc_parse=output-1.54.0/rustc-build/librustc_parse.rlib --extern rustc_plugin_impl=output-1.54.0/rustc-build/librustc_plugin_impl.rlib --extern rustc_save_analysis=output-1.54.0/rustc-build/librustc_save_analysis.rlib --extern rustc_codegen_ssa=output-1.54.0/rustc-build/librustc_codegen_ssa.rlib --extern rustc_session=output-1.54.0/rustc-build/librustc_session.rlib --extern rustc_error_codes=output-1.54.0/rustc-build/librustc_error_codes.rlib --extern rustc_interface=output-1.54.0/rustc-build/librustc_interface.rlib --extern rustc_serialize=output-1.54.0/rustc-build/librustc_serialize.rlib --extern rustc_ast=output-1.54.0/rustc-build/librustc_ast.rlib --extern rustc_span=output-1.54.0/rustc-build/librustc_span.rlib --extern rustc_mir_build=output-1.54.0/rustc-build/librustc_mir_build.rlib --extern rustc_typeck=output-1.54.0/rustc-build/librustc_typeck.rlib
BUILDING rustc_main from rustc-main v0.0.0 with features [llvm]
> /nix/store/54w5f18vb7xy2frx5i091q77l2zaz76x-mrustc-0.10/bin/mrustc rustc-1.54.0-src/compiler/rustc/src/main.rs -o output-1.54.0/rustc-build/rustc_main --crate-name rustc_main --crate-type bin -C emit-depfile=output-1.54.0/rustc-build/rustc_main.d --crate-tag 0_0_0_H2 -g --cfg debug_assertions -O -L output-1.54.0 -L output-1.54.0/rustc-build --cfg feature="llvm" --edition 2018 --extern rustc_driver=output-1.54.0/rustc-build/librustc_driver.rlib --extern rustc_codegen_ssa=output-1.54.0/rustc-build/librustc_codegen_ssa.rlib
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: warning: x86_64.o: missing .note.GNU-stack section implies executable stack
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: /nix/store/5i7p4ls89a87lw5sz7fydkg9xim4izxj-llvm-12.0.1-lib/lib/libLLVMX86CodeGen.a(X86WinEHState.cpp.o): in function `void std::deque<llvm::BasicBlock*, std::allocator<llvm::BasicBlock*> >::_M_push_back_aux<llvm::BasicBlock* const&>(llvm::BasicBlock* const&)':
(.text._ZNSt5dequeIPN4llvm10BasicBlockESaIS2_EE16_M_push_back_auxIJRKS2_EEEvDpOT_[_ZNSt5dequeIPN4llvm10BasicBlockESaIS2_EE16_M_push_back_auxIJRKS2_EEEvDpOT_]+0x207): undefined reference to `std::__throw_bad_array_new_length()'
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: /nix/store/5i7p4ls89a87lw5sz7fydkg9xim4izxj-llvm-12.0.1-lib/lib/libLLVMX86CodeGen.a(X86FloatingPoint.cpp.o): in function `(anonymous namespace)::FPS::runOnMachineFunction(llvm::MachineFunction&) [clone .part.0]':
(.text._ZN12_GLOBAL__N_13FPS20runOnMachineFunctionERN4llvm15MachineFunctionE.part.0+0xc49): undefined reference to `std::__throw_bad_array_new_length()'
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: /nix/store/5i7p4ls89a87lw5sz7fydkg9xim4izxj-llvm-12.0.1-lib/lib/libLLVMX86CodeGen.a(X86InstrInfo.cpp.o): in function `llvm::X86InstrInfo::getOutliningCandidateInfo(std::vector<llvm::outliner::Candidate, std::allocator<llvm::outliner::Candidate> >&) const':
(.text._ZNK4llvm12X86InstrInfo25getOutliningCandidateInfoERSt6vectorINS_8outliner9CandidateESaIS3_EE+0x8a6): undefined reference to `std::__throw_bad_array_new_length()'
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: /nix/store/5i7p4ls89a87lw5sz7fydkg9xim4izxj-llvm-12.0.1-lib/lib/libLLVMX86CodeGen.a(X86LoadValueInjectionLoadHardening.cpp.o): in function `(anonymous namespace)::X86LoadValueInjectionLoadHardeningPass::getGadgetGraph(llvm::MachineFunction&, llvm::MachineLoopInfo const&, llvm::MachineDominatorTree const&, llvm::MachineDominanceFrontier const&) const::{lambda(llvm::rdf::NodeAddr<llvm::rdf::DefNode*>)#2}::operator()(llvm::rdf::NodeAddr<llvm::rdf::DefNode*>) const::{lambda(llvm::rdf::NodeAddr<llvm::rdf::DefNode*>)#1}::operator()(llvm::rdf::NodeAddr<llvm::rdf::DefNode*>) const':
(.text._ZZZNK12_GLOBAL__N_138X86LoadValueInjectionLoadHardeningPass14getGadgetGraphERN4llvm15MachineFunctionERKNS1_15MachineLoopInfoERKNS1_20MachineDominatorTreeERKNS1_24MachineDominanceFrontierEENKUlNS1_3rdf8NodeAddrIPNSD_7DefNodeEEEE0_clESH_ENKUlSH_E_clESH_+0x1aba): undefined reference to `std::__throw_bad_array_new_length()'
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: /nix/store/5i7p4ls89a87lw5sz7fydkg9xim4izxj-llvm-12.0.1-lib/lib/libLLVMRISCVCodeGen.a(RISCVInstrInfo.cpp.o): in function `llvm::RISCVInstrInfo::getOutliningCandidateInfo(std::vector<llvm::outliner::Candidate, std::allocator<llvm::outliner::Candidate> >&) const':
(.text._ZNK4llvm14RISCVInstrInfo25getOutliningCandidateInfoERSt6vectorINS_8outliner9CandidateESaIS3_EE+0x72a): undefined reference to `std::__throw_bad_array_new_length()'
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: /nix/store/5i7p4ls89a87lw5sz7fydkg9xim4izxj-llvm-12.0.1-lib/lib/libLLVMPowerPCCodeGen.a(PPCInstrInfo.cpp.o):(.text._ZZNK4llvm12PPCInstrInfo28shouldReduceRegisterPressureEPNS_17MachineBasicBlockEPNS_17RegisterClassInfoEENKUlS2_E_clES2_+0x4a8): more undefined references to `std::__throw_bad_array_new_length()' follow
collect2: error: ld returned 1 exit status
C Compiler failed to execute - error code 256
Process exited with non-zero exit status 1
FAILING COMMAND:  /nix/store/54w5f18vb7xy2frx5i091q77l2zaz76x-mrustc-0.10/bin/mrustc rustc-1.54.0-src/compiler/rustc/src/main.rs -o output-1.54.0/rustc-build/rustc_main --crate-name rustc_main --crate-type bin -C emit-depfile=output-1.54.0/rustc-build/rustc_main.d --crate-tag 0_0_0_H2 -g --cfg debug_assertions -O -L output-1.54.0 -L output-1.54.0/rustc-build --cfg feature="llvm" --edition 2018 --extern rustc_driver=output-1.54.0/rustc-build/librustc_driver.rlib --extern rustc_codegen_ssa=output-1.54.0/rustc-build/librustc_codegen_ssa.rlib
BUILD FAILED
```

but it already failed with this error before this patch (though it did within 10 min instead of an hour), and I don't see how to fix it. Help welcome.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/mrustc/patches/0001-dont-download-rustc.patch [hunk 1]
```diff
@@ -0,0 +1,11 @@
+--- a/minicargo.mk	2023-04-06 08:26:18.408817572 +0200
++++ b/minicargo.mk	2023-04-06 08:27:11.553536996 +0200
+@@ -176,7 +176,7 @@
+ 	@echo [CURL] $@
+ 	@rm -f $@
+ 	@curl -sS https://static.rust-lang.org/dist/$@ -o $@
+-$(RUSTC_SRC_DL): $(RUSTC_SRC_TARBALL) rustc-$(RUSTC_VERSION)-src.patch
++$(RUSTC_SRC_DL): rustc-$(RUSTC_VERSION)-src.patch
```
Comment (progval) [inline pkgs/development/compilers/mrustc/patches/0001-dont-download-rustc.patch:8]: I don't get why this changes anything. The `$(RUSTC_SRC_DL)` file already exists, but for some reason Make wants to rebuild this target when it has `$(RUSTC_SRC_TARBALL)` as a dependency.

**Comment by r-burns** [comment]
That linker error is a libstdc++ version mismatch, fixed by removing the stdenv gcc 10 override in top-level. (I think this override was added by tooling during the gcc 10 -> 11 bump and not actually tested.)

```diff
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -15928,7 +15928,6 @@ with pkgs;
   mrustc = callPackage ../development/compilers/mrustc { };
   mrustc-minicargo = callPackage ../development/compilers/mrustc/minicargo.nix { };
   mrustc-bootstrap = callPackage ../development/compilers/mrustc/bootstrap.nix {
-    stdenv = gcc10StdenvCompat;
     openssl = openssl_1_1;
   };
 
```

That allows the rustc build by mrustc to succeed, but then mrustc fails to build cargo with an error when it encounters the `global_asm` macro. Interestingly, this is fixed by switching to an in-tree build of minicargo. I suppose there is some issue with the `--manifest-overrides` arg when using a system-installed minicargo, because any global_asm usage should be disabled by rustc-1.54.0-overrides.toml. (I'd like to look into this, but I suppose it's not a big deal if it gets rebuilt, since minicargo is very quick to compile and the vast majority of build time is spent on the actual rustc/cargo bootstrap stages.)

```diff
--- a/pkgs/development/compilers/mrustc/bootstrap.nix
+++ b/pkgs/development/compilers/mrustc/bootstrap.nix
@@ -72,7 +72,7 @@ stdenv.mkDerivation rec {
   makeFlags = [
     # Use shared mrustc/minicargo/llvm instead of rebuilding them
     "MRUSTC=${mrustc}/bin/mrustc"
-    "MINICARGO=${mrustc-minicargo}/bin/minicargo"
+    #"MINICARGO=${mrustc-minicargo}/bin/minicargo"
     "LLVM_CONFIG=${llvm_12.dev}/bin/llvm-config"
     "RUSTC_TARGET=${rust.toRustTarget stdenv.targetPlatform}"
   ];
@@ -89,6 +89,7 @@ stdenv.mkDerivation rec {
     export OUTDIR_SUF=-${rustcVersion}
     export RUSTC_VERSION=${rustcVersion}
     export MRUSTC_TARGET_VER=${mrustcTargetVersion}
+    export MRUSTC_PATH=${mrustc}/bin/mrustc
 
     echo minicargo.mk: libs
     make -f minicargo.mk "''${flagsArray[@]}" LIBS
```

After applying both of those changes to your PR, mrustc-bootstrap finished successfully.

**Comment by progval** [comment]
Thanks.

> (I'd like to look into this, but I suppose it's not a big deal if it gets rebuilt, since minicargo is very quick to compile and the vast majority of build time is spent on the actual rustc/cargo bootstrap stages.)

I tried this:

```diff
--- a/run_rustc/Makefile	2023-04-07 13:39:06.298958303 +0200
+++ b/run_rustc/Makefile	2023-04-07 13:40:53.476424274 +0200
@@ -152,9 +152,9 @@
 $(LIBDIR_S)libstd.rlib: $(BINDIR_S)rustc $(MINICARGO) Makefile
 	@echo "[MINICARGO] $(RUST_SRC_LIBS)std > $(LIBDIR_S)"
 	mkdir -p $(LIBDIR_S)
-	$V+MRUSTC_PATH=$(abspath $(BINDIR_S)rustc) $(MINICARGO) --vendor-dir $(VENDOR_DIR) --script-overrides ../script-overrides/stable-$(RUSTC_VERSION)-$(PLATFORM)/ --output-dir $(LIBDIR_S) $(RUST_SRC_LIBS)std
-	$V+MRUSTC_PATH=$(abspath $(BINDIR_S)rustc) $(MINICARGO) --vendor-dir $(VENDOR_DIR) --script-overrides ../script-overrides/stable-$(RUSTC_VERSION)-$(PLATFORM)/ --output-dir $(LIBDIR_S) $(RUST_SRC_LIBS)panic_unwind
-	$V+MRUSTC_PATH=$(abspath $(BINDIR_S)rustc) $(MINICARGO) --vendor-dir $(VENDOR_DIR) --script-overrides ../script-overrides/stable-$(RUSTC_VERSION)-$(PLATFORM)/ --output-dir $(LIBDIR_S) $(RUST_SRC_LIBS)test
+	$V+MRUSTC_PATH=$(abspath $(BINDIR_S)rustc) $(MINICARGO) --vendor-dir $(VENDOR_DIR) --script-overrides ../script-overrides/stable-$(RUSTC_VERSION)-$(PLATFORM)/ --manifest-overrides ../rustc-$(RUSTC_VERSION)-overrides.toml --output-dir $(LIBDIR_S) $(RUST_SRC_LIBS)std
+	$V+MRUSTC_PATH=$(abspath $(BINDIR_S)rustc) $(MINICARGO) --vendor-dir $(VENDOR_DIR) --script-overrides ../script-overrides/stable-$(RUSTC_VERSION)-$(PLATFORM)/ --manifest-overrides ../rustc-$(RUSTC_VERSION)-overrides.toml --output-dir $(LIBDIR_S) $(RUST_SRC_LIBS)panic_unwind
+	$V+MRUSTC_PATH=$(abspath $(BINDIR_S)rustc) $(MINICARGO) --vendor-dir $(VENDOR_DIR) --script-overrides ../script-overrides/stable-$(RUSTC_VERSION)-$(PLATFORM)/ --manifest-overrides ../rustc-$(RUSTC_VERSION)-overrides.toml --output-dir $(LIBDIR_S) $(RUST_SRC_LIBS)test
 
 # Build hello_world using the bootstrapping compiler (mrustc-built rustc)
 $(BINDIR_S)hello_world: $(RUST_SRC_HELLO) $(LIBDIR_S)libstd.rlib $(BINDIR_S)rustc

```

which fixes the issue (though I don't understand why it's only a problem when installing).

Now failing differently:

```
(18/19) BUILDING test v0.0.0 with features [std_detect_file_io,std_detect_dlsym_getauxval,panic-unwind]
> /nix/store/54w5f18vb7xy2frx5i091q77l2zaz76x-mrustc-0.10/bin/mrustc ../rustc-1.54.0-src/library/test/src/lib.rs -o output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libtest.rlib --crate-name test --crate-type rlib -C emit-depfile=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libtest.rlib.d --crate-tag 0_0_0_H640 -g --cfg debug_assertions -O -L output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib --cfg feature="std_detect_file_io" --cfg feature="std_detect_dlsym_getauxval" --cfg feature="panic-unwind" --edition 2018 --extern cfg_if=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-0_1_10_H3.rlib --extern getopts=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgetopts-0_2_21_H1.rlib --extern term=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libterm.rlib --extern std=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd.rlib --extern core=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore.rlib --extern libc=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-0_2_95_H19.rlib --extern panic_unwind=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind.rlib --extern panic_abort=output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_abort.rlib
output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libtest.rlib.c: In function 'ZRIG2cF10alloc0_0_04sync3Arc1gG4cD8std0_0_0_B4mpsc6stream6Packet1gG2cE14test0_0_0_H6405event13CompletedTest0g3new0g':
output-1.54.0/prefix-s/lib/rustlib/x86_64-unknown-linux-gnu/lib/libtest.rlib.c:69517:127: note: the ABI for passing parameters with 64-byte alignment has changed in GCC 4.6
69517 | static struct s_ZRG2cF10alloc0_0_04sync3Arc1gG4cD8std0_0_0_B4mpsc6stream6Packet1gG2cE14test0_0_0_H6405event13CompletedTest0g  ZRIG2cF10alloc0_0_04sync3Arc1gG4cD8std0_0_0_B4mpsc6stream6Packet1gG2cE14test0_0_0_H6405event13CompletedTest0g3new0g(
      |                                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error[E0463]: can't find crate for `std`

make: *** [Makefile:162: output-1.54.0/prefix-s/bin/hello_world] Aborted
```

I think I'll do it your way

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 224970` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>mrustc</li>
    <li>mrustc-bootstrap</li>
    <li>mrustc-minicargo</li>
  </ul>
</details>

File: pkgs/development/compilers/mrustc/bootstrap.nix [hunk 2]
```diff
@@ -78,8 +72,8 @@ stdenv.mkDerivation rec {
   makeFlags = [
     # Use shared mrustc/minicargo/llvm instead of rebuilding them
     "MRUSTC=${mrustc}/bin/mrustc"
-    "MINICARGO=${mrustc-minicargo}/bin/minicargo"
-    "LLVM_CONFIG=${llvm_7.dev}/bin/llvm-config"
+    #"MINICARGO=${mrustc-minicargo}/bin/minicargo"  # FIXME: uncomment this so minicargo doesn't get rebuilt
```
Comment (RaitoBezarius) [inline pkgs/development/compilers/mrustc/bootstrap.nix:75]: Do we want to uncomment it or not?

**Comment by RaitoBezarius** [comment]
Thank you for the PR, @progval ; Can I ask you to reformat your "Remote obsolete patches" commit like the others? I think this is good to go.

File: pkgs/development/compilers/mrustc/bootstrap.nix (hunk 2 above)
Comment (progval) [reply pkgs/development/compilers/mrustc/bootstrap.nix:75]: nah, couldn't find how to make it work

**Comment by progval** [comment]
Don't you want to squash the PR?

**Comment by RaitoBezarius** [comment]
> Don't you want to squash the PR?

No, all your commit seems properly atomic IMHO. They can stay as-is.

**Comment by progval** [comment]
done

File: pkgs/development/compilers/mrustc/bootstrap.nix (hunk 2 above)
Comment (RaitoBezarius) [reply pkgs/development/compilers/mrustc/bootstrap.nix:75]: we should update the comment or get rid of it so that people can understand the difficulty of the TODO/FIXME

File: pkgs/development/compilers/mrustc/patches/0001-dont-download-rustc.patch (hunk 1 above)
Comment (RaitoBezarius) [reply pkgs/development/compilers/mrustc/patches/0001-dont-download-rustc.patch:8]: Ideally this comment/information should end in the header of this patch somewhat

**Comment by RaitoBezarius** [comment]
Thousands apologies again for my mistake thinking the last comment was not addressed. In the future, please spam me by all means if I forget something and do not hesitate to ask me for review/unblocking PRs. :)
